benchmarks/benchmark_utils.py

```diff
@@ -0,0 +1,297 @@
+import os
+import pandas as pd
+import numpy as np
+from pyspark.ml.recommendation import ALS
+import pyspark.sql.functions as F
+from pyspark.sql import SparkSession
+from pyspark.sql.types import StructType, StructField
+from pyspark.sql.types import StringType, FloatType, IntegerType, LongType
+from fastai.collab import EmbeddingDotBias, collab_learner, CollabDataBunch, load_learner
+import surprise
+
+from reco_utils.common.general_utils import get_number_processors
+from reco_utils.common.timer import Timer
+from reco_utils.common.gpu_utils import get_cuda_version, get_cudnn_version
+from reco_utils.common.spark_utils import start_or_get_spark
+from reco_utils.dataset import movielens
+from reco_utils.dataset.sparse import AffinityMatrix
+from reco_utils.dataset.python_splitters import python_chrono_split
+from reco_utils.recommender.sar.sar_singlenode import SARSingleNode
+from reco_utils.recommender.ncf.ncf_singlenode import NCF
+from reco_utils.recommender.ncf.dataset import Dataset as NCFDataset
+from reco_utils.recommender.surprise.surprise_utils import surprise_trainset_to_df
+from reco_utils.recommender.fastai.fastai_utils import hide_fastai_progress_bar, cartesian_product, score
+from reco_utils.evaluation.spark_evaluation import SparkRatingEvaluation, SparkRankingEvaluation
+from reco_utils.evaluation.python_evaluation import map_at_k, ndcg_at_k, precision_at_k, recall_at_k
+from reco_utils.evaluation.python_evaluation import rmse, mae, rsquared, exp_var
+
+
+TOP_K = 10
+USER_COL = "UserId"
+ITEM_COL = "MovieId"
+RATING_COL = "Rating"
+TIMESTAMP_COL = "Timestamp"
+PREDICTION_COL = "prediction"
+SEED = 77
+
+
+def prepare_training_als(train):
+    schema = StructType(
+    (
+        StructField(USER_COL, IntegerType()),
+        StructField(ITEM_COL, IntegerType()),
+        StructField(RATING_COL, FloatType()),
+        StructField(TIMESTAMP_COL, LongType()),
+    )
+    )
+    spark = start_or_get_spark()
+    return spark.createDataFrame(train, schema)
+
+
+def train_als(params, data):
+    symbol = ALS(**params)
+    with Timer() as t:
+        model = symbol.fit(data)
+    return model, t
+
+
+def prepare_metrics_als(train, test):
+    schema = StructType(
+    (
+        StructField(USER_COL, IntegerType()),
+        StructField(ITEM_COL, IntegerType()),
+        StructField(RATING_COL, FloatType()),
+        StructField(TIMESTAMP_COL, LongType()),
+    )
+    )
+    spark = start_or_get_spark()
+    return prepare_training_als(train), spark.createDataFrame(test, schema)
+
+
+def predict_als(model, test):
+    with Timer() as t:
+        preds = model.transform(test)
+    return preds, t
+
+
+def recommend_k_als(model, test, train):
+    with Timer() as t:
+        # Get the cross join of all user-item pairs and score them.
+        users = train.select(USER_COL).distinct()
+        items = train.select(ITEM_COL).distinct()
+        user_item = users.crossJoin(items)
+        dfs_pred = model.transform(user_item)

```

Comment:
wouldn't it be better to filter out seen items before prediction?

also it would be awesome to extract thisout as a function that can be used by all algorithms: get_all_user_items(data, user_col, item_col, remove_seen=True)

Reply:
Moving this to a future issue since there was a similar ask from Le https://github.com/Microsoft/Recommenders/issues/733